From 1972a3a0eafc718817cfe995baa12c6faddc19dd Mon Sep 17 00:00:00 2001
From: Andrei Varvara <andrei.varvara@freescale.com>
Date: Fri, 22 Nov 2013 17:26:20 +0200
Subject: [PATCH 362/429] dpa_offload: Bug fix, inbound flow id was not
 released properly

Properly release the inbound flow id in rollback and
in case or SA removal

Signed-off-by: Andrei Varvara <andrei.varvara@freescale.com>
Change-Id: If319b1498f3541c5aba235d3fbc95dff216c3e64
Reviewed-on: http://git.am.freescale.net:8181/6829
Tested-by: Review Code-CDREVIEW <CDREVIEW@freescale.com>
Reviewed-by: Thomas Trefny <Tom.Trefny@freescale.com>
[origin patch is from QorIQ-SDK-V1.5-20131219-yocto]
Signed-off-by: Yang Wei <Wei.Yang@windriver.com>
---
 drivers/staging/fsl_dpa_offload/dpa_ipsec.c |   24 ++++++++++++------------
 1 files changed, 12 insertions(+), 12 deletions(-)

diff --git a/drivers/staging/fsl_dpa_offload/dpa_ipsec.c b/drivers/staging/fsl_dpa_offload/dpa_ipsec.c
index c6d1866..739fe89 100644
--- a/drivers/staging/fsl_dpa_offload/dpa_ipsec.c
+++ b/drivers/staging/fsl_dpa_offload/dpa_ipsec.c
@@ -3129,6 +3129,17 @@ static int put_sa(struct dpa_ipsec_sa *sa)
 		return -EDOM;
 	}
 
+	/* Release the flow ID - only for inbound SAs */
+	if (sa_is_inbound(sa) && !ignore_post_ipsec_action(dpa_ipsec) &&
+	    sa->inbound_flowid != INVALID_INB_FLOW_ID) {
+		err = put_inbound_flowid(dpa_ipsec, sa->inbound_flowid);
+		if (err < 0) {
+			log_err("Could not put flow id in circular queue.\n");
+			return err;
+		}
+		sa->inbound_flowid = INVALID_INB_FLOW_ID;
+	}
+
 	/* Mark as free index in used SA IDs vector of this DPA IPSEC instance*/
 	dpa_ipsec->used_sa_ids[sa->used_sa_index] = DPA_OFFLD_INVALID_OBJECT_ID;
 	dpa_ipsec->num_used_sas--;
@@ -3189,18 +3200,7 @@ remove_fq_pair:
 		return err_rb;
 	}
 
-	/* Free the SA id and FlowID (for inbound SAs only).*/
-	if (sa_is_inbound(sa) &&
-	    !ignore_post_ipsec_action(dpa_ipsec) &&
-	    sa->inbound_flowid != INVALID_INB_FLOW_ID) {
-		err_rb = put_inbound_flowid(dpa_ipsec, sa->inbound_flowid);
-		if (err_rb < 0) {
-			log_err("Could not put flow id in circular queue.\n");
-			return err_rb;
-		}
-		sa->inbound_flowid = INVALID_INB_FLOW_ID;
-	}
-
+	/* Release the SA */
 	err_rb = put_sa(sa);
 
 	return err_rb;
-- 
1.7.5.4

